src/WinRPM.jl

```diff
@@ -46,33 +46,55 @@ function __init__()
     update(false, false)
 end
 
-if isunix()
-    function download(source::AbstractString)
-        x = HTTPC.get(source)
-        unsafe_string(x.body), x.http_code
+if VERSION < v"0.7.0-DEV"
+    if isunix()
+        function download(source::AbstractString)
+            x = HTTPC.get(source)
+            unsafe_string(x.body), x.http_code
+        end
+    elseif iswindows()
+        function download(source::AbstractString; retry=5)
+            dest = Vector{UInt16}(261)
+            for i in 1:retry
+                res = ccall((:URLDownloadToCacheFileW, :urlmon), stdcall, Cuint,
+                (Ptr{Void}, Ptr{UInt16}, Ptr{UInt16}, Clong, Cint, Ptr{Void}),
+                C_NULL, transcode(UInt16, source), dest, sizeof(dest) >> 1, 0, C_NULL)
+                if res == 0
+                    resize!(dest, findfirst(iszero, dest) - 1)
+                    filename = transcode(String, dest)
+                    if isfile(filename)
+                        return read(filename, String), 200
+                    end
+                else
+                    warn("Unknown download failure, error code: $res")
+                end
+                warn("Retry $i/$retry downloading: $source")
+                return "", 0
+            end
+        end
+    else
+        error("Platform not supported: $(Sys.KERNEL)")
     end
-elseif iswindows()
+else
     function download(source::AbstractString; retry=5)
-        dest = Vector{UInt16}(261)
         for i in 1:retry
-            res = ccall((:URLDownloadToCacheFileW, :urlmon), stdcall, Cuint,
-              (Ptr{Void}, Ptr{UInt16}, Ptr{UInt16}, Clong, Cint, Ptr{Void}),
-              C_NULL, transcode(UInt16, source), dest, sizeof(dest) >> 1, 0, C_NULL)
-            if res == 0
-                resize!(dest, findfirst(iszero, dest) - 1)
-                filename = transcode(String, dest)
+            try
+                filename = joinpath(tempdir(), split(source, "/")[end])
+                filename = Base.download(source, filename)
                 if isfile(filename)
-                    return read(filename, String), 200
+                    return readstring(filename), 200
                 end
-            else
-                warn("Unknown download failure, error code: $res")
+            catch ex
+                warn("Unknown download failure. Retry $i/$retry downloading: $source")
             end
-            warn("Retry $i/$retry downloading: $source")
         end
+        warn("""Unknown download failure in `Base.download` function""")
+        if iswindows()
```

Comment:
Can we just rethrow the exception raised by `Base.download`? It needs to print a more helpful error message anyway (e.g. it could check what PowerShell version is installed), so no need to duplicate that logic.

Reply:
We shouldn't rethrow the error. This is not what the rest of WinRPM code expect - it expects to print the error as a warning and send a errcode !=200. The `WinRPM.update` function then silently recovers and tries a different provider. 
* I'd suggest catching an error thrown by `Base.download`, display it as a warning and continue the same path the old code expects. And we should update the `Base.download` function to give more info in its error exception.
* Or, we can rewrite also the other parts of WinRPM - but that won't bring much functionality either.